File: frontend/lib/models/alert.dart
import 'package:flutter/foundation.dart';

@immutable
class Alert {
  final String id;
  final String parcelId;
  final String parcelName;
  final String alertType;
  final String severity;
  final String title;
  final String description;
  final String? aiRecommendation;
  final bool isRead;
  final DateTime createdAt;
  final DateTime? readAt;
  final double? triggeredValue;
  final double? thresholdValue;

  const Alert({
    required this.id,
    required this.parcelId,
    required this.parcelName,
    required this.alertType,
    required this.severity,
    required this.title,
    required this.description,
    this.aiRecommendation,
    required this.isRead,
    required this.createdAt,
    this.readAt,
    this.triggeredValue,
    this.thresholdValue,
  });

  factory Alert.fromJson(Map<String, dynamic> json) {
    return Alert(
      id: json['id'] as String,
      parcelId: json['parcel_id'] as String,
      parcelName: json['parcel_name'] as String,
      alertType: json['alert_type'] as String,
      severity: json['severity'] as String,
      title: json['title'] as String,
      description: json['description'] as String,
      aiRecommendation: json['ai_recommendation'] as String?,
      isRead: json['is_read'] as bool,
      createdAt: DateTime.parse(json['created_at'] as String),
      readAt: json['read_at'] != null
          ? DateTime.parse(json['read_at'] as String)
          : null,
      triggeredValue: (json['triggered_value'] as num?)?.toDouble(),
      thresholdValue: (json['threshold_value'] as num?)?.toDouble(),
    );
  }

  bool get isCritical => severity == 'critical' || severity == 'high';
  bool get hasRecommendation => aiRecommendation != null && aiRecommendation!.isNotEmpty;
}

File: backend/app/presentation/schemas/alert.py
"""Pydantic schemas for the Alerts endpoints."""

from __future__ import annotations

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel


class AlertResponse(BaseModel):
    """Serialized alert for the mobile app."""

    id: UUID
    parcel_id: UUID
    parcel_name: str
    alert_type: str
    severity: str
    title: str
    description: str
    ai_recommendation: Optional[str] = None
    is_read: bool
    created_at: datetime
    read_at: Optional[datetime] = None
    triggered_value: Optional[float] = None
    threshold_value: Optional[float] = None

    model_config = {"from_attributes": True}


class UnreadCountResponse(BaseModel):
    """Badge count response."""

    count: int
